test/functional/blocknotify.py

```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env python3
+# Copyright (c) 2017 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+"""Test blocknotify.
+
+Verify that a bitcoind node calls the given command for each block
+"""
+
+import os
+import tempfile
+import time
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal
+
+class BlockNotifyTest(BitcoinTestFramework):
+
+    def __init__(self):
+        super().__init__()
+
+        # get a temporary file path to store the notified blocks
+        fd, path = tempfile.mkstemp(text = True)
+        os.close(fd)
+
+        self.path = path
+        self.setup_clean_chain = False
+        self.num_nodes = 1
+
+        # the command will append the block hash to file
+        self.extra_args = [['-blocknotify=echo -n %%s >> %s' % path]]
+
+    def run_test(self):
+        # generate 3 blocks
+        blocks = self.nodes[0].generate(3)
+
+        # wait while the command runs asynchronously
+        time.sleep(3)
```

Comment:
Please no fixed sleeps. This might be too much time on a fast system or too less time on a slow io system.

You could wrap it into a while loop with timeout on failure and break on success. Alternatively there should be some helper function `wait_until` if I am not mistaken.